lib/api/messages.js

```diff
@@ -1538,6 +1538,206 @@ module.exports = (db, server, messageHandler, userHandler, storageHandler) => {
         })
     );
 
+    /**
+     * @api {post} /users/:user/mailboxes/:mailbox/messages/:message/snooze Snooze a message for later
+     * @apiName SnoozeMessage
+     * @apiGroup Messages
+     * @apiHeader {String} X-Access-Token Optional access token if authentication is enabled
+     * @apiHeaderExample {json} Header-Example:
+     * {
+     *   "X-Access-Token": "59fc66a03e54454869460e45"
+     * }
+     *
+     * @apiParam {String} user ID of the User
+     * @apiParam {String} mailbox ID of the Mailbox
+     * @apiParam {Number} message ID of the Message
+     * @apiParam {Number} The time duration (in milliseconds) to keep the message snoozed for
```

Comment:
This is not currently used in any way, does not even have a name defined.

It would probably make sense to store it as an additional field in the message object by the `moveMessage` method, otherwise it would be difficult to say when would snoozing end for a message. The same field should be exposed in the output of https://api.wildduck.email/#api-Messages-GetMessage (actual snooze move should be handled by not-yet-existing snooze task and it should not rely on the snooze information stored to the message object as it would be for informational purposes only)

As we only want to have snooze information for messages added to the Snooze folder by the API call and not messages that are copied around by the user, then `messageHandler.move` and `messageHandler.add`  should set snooze information to `false` unless snooze information is provided. 

Normal message move:
```
messageHandler.move({some_data})
// messageObject.snoozeTime = false;
```

Snooze move:
```
messageHandler.move({some_data, snoozeTime: new Date(Date.now()+1000)})
// messageObject.snoozeTime = options.snoozeTime;
```

Reply:
So my idea was to use the snooze time from the API call, add the current time to it (`Date.now() + snoozeTime` since that way, we'll avoid any timezone issues rather than having the client send the time to unsnooze), store the time (in a separate object? hadn't gotten through to that part yet) and then  either use `setTimeout` (not the best way, I know) or `node-schedule` to schedule the unsnoozing part.

But I'll take a look at the `moveMessage` code and see how to add the snoozing part to the message object, thanks.

As for the tasks part, I'm confused as to what exactly a task is and what it's meant for. Could you help me understand please?

Reply:
You could store an information object instead of end time as that entire data would be informational only, does not really matter what you store as long as it's usable information.

```
let now = new Date();
messageObject.snooze = {
  "initiated": now,
  "ends": new Date(now.getTime() + duration),
  "duration": duration
}
```

The unsnoozing part should be handled by internal task system. Unfortunately it does not support delays yet but this can be fixed. Currently tasks are added diretly to the tasks collection in db, like [here](https://github.com/nodemailer/wildduck/blob/master/lib/api/users.js#L1887-L1893) (in the future there should probably be some kind of wrapper that normalises db entries.):

```
await db.database.collection('tasks').insertOne({
    task: 'unsnooze',
    locked: false,
    lockedUntil: now,
    created: now,
    status: 'queued',
    runAfter: new Date(Date.now() + duration) // this part is not implemented yet
});
```

The actual unsnoozing task can wait until there are messages queued for unsnoozing.